### PR TITLE
feat(merkle-tree): Add canopy

### PR DESCRIPTION
**Pull request by vadorovsky**
Canopy is a cache of N uppermost nodes in the Merkle tree. The purpose of it is to allow clients to send smaller Merkle proofs, which don't include the N last nodes. The purpose is to minimize the instruction data size.

### CODE REVIEW

File: merkle-tree/concurrent/src/lib.rs [hunk 1]
```diff
@@ -59,8 +59,7 @@ where
     pub roots_length: usize,
     pub current_root_index: usize,
 
-    // TODO(vadorovsky): Add canopy.
-    pub canopy_depth: usize,
+    pub canopy_depth: u32,
```
Comment (ananas-block) [inline merkle-tree/concurrent/src/lib.rs:62]: even u8 would be enough right?
It's fighting over scraps just saying.

Comment (vadorovsky) [reply merkle-tree/concurrent/src/lib.rs:62]: I picked u32, because all `.pow()` implementations expect exponential to be u32, e.g.

https://doc.rust-lang.org/std/primitive.usize.html#method.pow

And we are using pow in `canopy_size` method (basically `2_usize.pow(canopy_depth - 1) - 1`) to figure out the number of nodes in canopy.

File: programs/account-compression/src/state/public_state_merkle_tree.rs [hunk 2]
```diff
@@ -126,13 +132,14 @@ mod test {
...
-            state_merkle_tree_struct: [0u8; 224],
+            state_merkle_tree_struct: [0u8; 256],
             state_merkle_tree_filled_subtrees: [0u8; 832],
             state_merkle_tree_changelog: [0u8; 1220800],
             state_merkle_tree_roots: [0u8; 76800],
+            state_merkle_tree_canopy: [0u8; 65472],
         };
 
-        let merkle_tree = account.load_merkle_tree_init(26, 1400, 2400).unwrap();
+        let merkle_tree = account.load_merkle_tree_init(26, 1400, 2400, 10).unwrap();
```
Comment (ananas-block) [inline programs/account-compression/src/state/public_state_merkle_tree.rs:142]: could we use the constants here?

File: merkle-tree/concurrent/tests/tests.rs [hunk 3]
```diff
@@ -634,49 +636,125 @@ where
     }
 }
 
+fn compat_canopy<H, const HEIGHT: usize>()
```
Comment (ananas-block) [inline merkle-tree/concurrent/tests/tests.rs:639]: do you think it could make sense to add a long running test that tries height 26 with canopy 10, to explicitly test the parameters we are going to use onchain?
Another interesting testcase could be comparing a tree of height 10 and canopy zero with a  tree of height 10 with canopy 10 and fill it up to assert that the canopy behaves exactly the same for every leaf.

Comment (ananas-block) [reply merkle-tree/concurrent/tests/tests.rs:639]: Can you think of other edge cases to test?

**Review by ananas-block — COMMENTED** [review]
Thanks for the pr, excited for the canopies!
Some minor things and a couple ideas for additional tests.

File: merkle-tree/concurrent/tests/tests.rs (hunk 3 above)
Comment (vadorovsky) [reply merkle-tree/concurrent/tests/tests.rs:639]: I can give it more thought, but the only other scenario which comes immediately to my mind is testing indexed MT with canopy (appends followed by update).

File: programs/account-compression/src/state/public_state_merkle_tree.rs (hunk 2 above)
Comment (vadorovsky) [reply programs/account-compression/src/state/public_state_merkle_tree.rs]: done

**Review by ananas-block — APPROVED** [review]
